private/requestcompression/request_compression.go

```diff
@@ -0,0 +1,137 @@
+const (
+	GZIP Algorithm = "gzip"
+)
+
+type compressFunc func(io.Reader) ([]byte, error)
+
+var allowedAlgorithms = map[Algorithm]compressFunc{
+	GZIP: gzipCompress,
+}
+
+// AddRequestCompression add requestCompression middleware to op stack
+func AddRequestCompression(stack *middleware.Stack, DisableRequestCompression bool, RequestMinCompressSizeBytes int64) error {
+	return stack.Serialize.Add(&requestCompression{
+		disableRequestCompression:   DisableRequestCompression,
+		requestMinCompressSizeBytes: RequestMinCompressSizeBytes,
+		compressAlgorithms:          []Algorithm{GZIP},
+	}, middleware.After)
+}
+
+type requestCompression struct {
+	disableRequestCompression   bool
+	requestMinCompressSizeBytes int64
+	compressAlgorithms          []Algorithm
+}
+
+// ID returns the ID of the middleware
+func (m requestCompression) ID() string {
+	return "RequestCompression"
+}
+
+// HandleSerialize gzip compress the request's stream/body if enabled by config fields
+func (m requestCompression) HandleSerialize(
+	ctx context.Context, in middleware.SerializeInput, next middleware.SerializeHandler,
+) (
+	out middleware.SerializeOutput, metadata middleware.Metadata, err error,
+) {
+	if m.disableRequestCompression {
+		return next.HandleSerialize(ctx, in)
+	}
+	// still need to check requestMinCompressSizeBytes in case it is out of range after service client config
+	if m.requestMinCompressSizeBytes < 0 || m.requestMinCompressSizeBytes > maxRequestMinCompressSizeBytes {
+		return out, metadata, fmt.Errorf("invalid range for min request compression size bytes %d, must be within 0 and 10485760 inclusively", m.requestMinCompressSizeBytes)
+	}
+
+	req, ok := in.Request.(*http.Request)
+	if !ok {
+		return out, metadata, fmt.Errorf("unknown request type %T", req)
+	}
+
+	for _, algorithm := range m.compressAlgorithms {
+		compressFunc := allowedAlgorithms[algorithm]
+		if compressFunc != nil {
+			if stream := req.GetStream(); stream != nil {
+				compressedBytes, err := compressFunc(stream)
+				if err != nil {
+					return out, metadata, fmt.Errorf("failed to compress request stream, %v", err)
+				}
+
+				var newReq *http.Request
+				if newReq, err = req.SetStream(bytes.NewReader(compressedBytes)); err != nil {
+					return out, metadata, fmt.Errorf("failed to set request stream, %v", err)
+				}
+				*req = *newReq
+			} else {
+				if req.Body == nil {
+					break
+				}
+				body, err := io.ReadAll(req.Body)
+				if err != nil {
+					return out, metadata, fmt.Errorf("failed to read request body")
+				}
+				if int64(len(body)) < m.requestMinCompressSizeBytes {
+					req.Body = io.NopCloser(bytes.NewReader(body))
+					break
+				}
+				compressedBytes, err := compressFunc(bytes.NewReader(body))
+				if err != nil {
+					return out, metadata, fmt.Errorf("failed to compress request body, %v", err)
+				}
+
+				req.Body = io.NopCloser(bytes.NewReader(compressedBytes))
+			}
+
+			req.Header.Add("Content-Encoding", "gzip")
```

Comment:
Shouldn't you take the name of the selected algorithm instead of hardcoding gzip?